packages/engine/src/config/topology.rs

```diff
@@ -164,287 +150,471 @@ pub enum WrappingBehavior {
+    pub move_wrapped_agents: bool,
 
-            if config.wrap_y_mode == WrappingBehavior::OffsetReflection {
-                config.wrap_x_mode = WrappingBehavior::Continuous;
-            }
-            if config.wrap_z_mode == WrappingBehavior::OffsetReflection {
-                config.wrap_y_mode = WrappingBehavior::Continuous;
-            }
+    /// Cache how many wrapped points we need to calculate
+    pub wrapping_combinations: usize,
+}
 
-            let wrapping_combinations =
-                vec![config.wrap_x_mode, config.wrap_y_mode, config.wrap_z_mode]
-                    .iter()
-                    .fold(1, |acc, wrap| -> usize {
-                        if let WrappingBehavior::NoWrap = wrap {
-                            acc
-                        } else {
-                            acc * 2
-                        }
-                    });
-            config.wrapping_combinations = wrapping_combinations;
-
-            if let Some(search_radius) = topology_props
-                .get("search_radius")
-                .and_then(serde_json::Value::as_f64)
-            {
-                config.search_radius = Some(search_radius);
-            }
+impl Default for Config {
+    fn default() -> Config {
+        Config {
+            bounds: Default::default(),
+            wrap_modes: Default::default(),
+            search_radius: None,
+            distance_function: DistanceFunction::default().as_function(),
+            move_wrapped_agents: true,
+            wrapping_combinations: 1,
+        }
+    }
+}
 
-            if let Some(serde_json::Value::String(distance_func_str)) =
-                topology_props.get("distance_function")
-            {
-                config.distance_function = Box::new(match &**distance_func_str {
-                    "manhattan" => manhattan,
-                    "euclidean" => euclidean,
-                    "euclidean_squared" => euclidean_squared,
-                    // "conway" | "chebyshev" => conway,
-                    // default is already conway, TODO nicer alternative
-                    _ => conway,
-                });
-            }
+impl Config {
+    /// Creates a topology configuration from [`Globals`].
+    ///
+    /// If a value is not set, it will fallback as specified in [`Config::default()`].
+    ///
+    /// # Panics
+    ///
+    /// - if x-wrapping behavior is [`WrappingBehavior::OffsetReflection`]
+    /// - if y-wrapping behavior **and** z-wrapping behavior is
+    ///   [`WrappingBehavior::OffsetReflection`]
+    pub fn from_globals(globals: &Globals) -> Self {
+        fn from_json<T>(topology: &mut serde_json::Map<String, Value>, key: &str, fallback: T) -> T
+        where
+            T: Serialize + for<'de> Deserialize<'de>,
+        {
+            topology
+                .remove(key)
+                .and_then(|value| {
+                    serde_json::from_value(value)
```

Comment:
I feel like if a user inputs an invalid value we should error rather than falling back to the default